satpy/readers/amsr2_l2_gaasp.py

```diff
@@ -0,0 +1,286 @@
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""GCOM-W1 AMSR2 Level 2 files from the GAASP software.
+
+GAASP output files are in the NetCDF4 format. Software is provided by NOAA
+and is also distributed by the CSPP group. More information on the products
+supported by this reader can be found here:
+https://www.star.nesdis.noaa.gov/jpss/gcom.php for more information.
+
+GAASP includes both swath/granule products and gridded products. Swath
+products are provided in files with "MBT", "OCEAN", "SNOW", or "SOIL" in the
+filename. Gridded products are in files with "SEAICE-SH" or "SEAICE-NH" in the
+filename where SH stands for South Hemisphere and NH stands for North
+Hemisphere. These gridded products are on the EASE2 North pole and South pole
+grids. See https://nsidc.org/ease/ease-grid-projection-gt for more details.
+
+Note that since SEAICE products can be on both the northern or
+southern hemisphere or both depending on what files are provided to Satpy, this
+reader appends a `_NH` and `_SH` suffix to all variable names that are
+dynamically discovered from the provided files.
+
+"""
+
+import logging
+from datetime import datetime
+
+import numpy as np
+import xarray as xr
+from pyproj import CRS
+from pyresample.geometry import AreaDefinition
+
+from satpy.readers.file_handlers import BaseFileHandler
+from satpy import CHUNK_SIZE
+
+logger = logging.getLogger(__name__)
+
+
+class GAASPFileHandler(BaseFileHandler):
+    """Generic file handler for GAASP output files."""
+
+    y_dims = (
+        'Number_of_Scans',
+    )
+    x_dims = (
+        'Number_of_hi_rez_FOVs',
+        'Number_of_low_rez_FOVs',
+    )
+    time_dims = (
+        'Time_Dimension',
+    )
+    is_gridded = False
+    dim_resolutions = {
+        'Number_of_hi_rez_FOVs': 5000,
+        'Number_of_low_rez_FOVs': 10000,
+    }
+
+    def __init__(self, filename, filename_info, filetype_info):
+        """Open the NetCDF file with xarray and prepare the Dataset for reading."""
+        super().__init__(filename, filename_info, filetype_info)
+        chunks = {dim_name: CHUNK_SIZE for dim_name in
+                  self.y_dims + self.x_dims + self.time_dims}
+        self.nc = xr.open_dataset(self.filename,
+                                  decode_cf=True,
+                                  mask_and_scale=False,
+                                  chunks=chunks)
+
+        if len(self.time_dims) == 1:
+            self.nc = self.nc.rename({self.time_dims[0]: 'time'})
+
+    @property
+    def start_time(self):
+        """Get start time of observation."""
+        time_str = self.nc.attrs['time_coverage_start']
+        return datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+    @property
+    def end_time(self):
+        """Get end time of observation."""
+        time_str = self.nc.attrs['time_coverage_end']
+        return datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%fZ")
```

Comment:
Looks like the start time and end time is provided in the filename. Could these be checked first, and go into reading the nc values if the filename times aren't available?

Reply:
In my experience the start/end time in the filenames doesn't always match the internal data and it is best to use the file content first. What would be the benefit of using the filename times?

Reply:
That would be the time the user expects based on the filename. I think the internal time could be confusing.

Reply:
I would assume the internal time is more accurate though. I have also had cases where users expect the same time as in the filename but that doesn't make them right.

Reply:
I recently noticed that reading the value inside the file takes a significant amount of time when having the files on a shared filesystem over a DSL Internet connection. That happened in a case where the files were provided to satpy but ended not being loaded because the data was elsewhere (viirs sdr reader). With the advance of remote file systems like s3 buckets, I suspect this will be all the more common situation. So, if the times are similar in 99% of the cases, I'd rather use the filename time.
This maybe hints for having two kinds of times in the future, the filename ones (rough estimate) for basic filtering of the data, and a more precise one for applications that need it. But that is another issue.

Reply:
I think this depends on the format being read. I'd have to check, but I have always assumed that xarray.open_dataset with a remote file will always load at least the global attributes. I suppose in this case since start and end are both in the filename I'd be ok doing this though.